### DISCUSSION TITLE
Store array variable

### DISCUSSION THREAD

**Discussion by r0bchain** [Cyfrin/foundry-full-course-cu#1939] [General]
Hi there! As a smart contract beginner, I have many questions coming to mind these days. One of them is about the cost of increasing an array and who pays for this extra storage. For example, in the Ruffle example from this great course, if the s_players[] array starts with only one user (index 0) but grows over time (every time a new player joins the game), the price of each transaction will increase as well, right? This is because the array becomes larger and requires more storage.

If this is true, another question, more ethical than technical, comes to mind. Is it fair that the first user pays X and subsequent users pay X + N (for the same "service"), where N is always increasing?

Are there better solutions for this price increase issue? A possible solution that comes to mind is to "buy" a fixed array size, paid for by the owner of the contract during its creation. This way, all users would pay the same. Maybe I'm saying something impractical, but I'm curious about the feasibility of this approach.

Thanks.

**Comment by EngrPips** [comment]
Hello @r0bertinski, an array indeed grows bigger with every element we push into it, and the bigger it grows, the more expensive the operations we do on the array. The first player would pay less to be added to the array than the 10th player, and that's how it works for an array in solidity. It seems unfair that one person is paying far more to access the same service, but even life is unfair.

You mentioned that the contract owner buys an array of fixed sizes, so everyone doesn't need to pay to enter the array. Technically, the user cannot buy an array, but the owner of the contract can sponsor every transaction that is sent to the contract by using AA(Account Abstraction) link to read more about it `https://medium.com/coinmonks/account-abstraction-a-tool-against-gas-fees-in-your-dapp-322d66670395`. That way no one pays any gas fee to enter the raffle.

But Looking at our raffle contract and its purpose, you would see that it doesn't make sense for the contract owner to cover gas fees for users as they are meant to enter into a lottery to win a price. So sometimes, some people pay more for gas than others. There could be some solution where we won't use an array at all and avoid all those iterations altogether, but I am scared it can be a pretty complex solution.

**Reply by Robschuh** [reply]
The sponsor solution could be great for some situations, possible (as you said), not for the ruffle example.
Thanks 🙏

**Reply by EngrPips** [reply]
Exactly.
